pkg/query-service/app/preferences/model.go

```diff
@@ -0,0 +1,479 @@
+package preferences
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/jmoiron/sqlx"
+	"go.signoz.io/signoz/ee/query-service/model"
+)
+
+type Range struct {
+	Min int64 `json:"min"`
+	Max int64 `json:"max"`
+}
+
+type Preference struct {
+	Key              string        `json:"key"`
+	Name             string        `json:"name"`
+	Description      string        `json:"description"`
+	ValueType        string        `json:"valueType"`
+	DefaultValue     interface{}   `json:"defaultValue"`
+	AllowedValues    []interface{} `json:"allowedValues"`
+	IsDiscreteValues bool          `json:"isDiscreteValues"`
+	Range            Range         `json:"range"`
+	AllowedScopes    []string      `json:"allowedScopes"`
+}
+
+func (p *Preference) ErrorValueTypeMismatch() *model.ApiError {
+	return &model.ApiError{Typ: model.ErrorBadData, Err: fmt.Errorf("the preference value is not of expected type: %s", p.ValueType)}
+}
+
+const (
+	PreferenceValueTypeInteger string = "integer"
+	PreferenceValueTypeFloat   string = "float"
+	PreferenceValueTypeString  string = "string"
+	PreferenceValueTypeBoolean string = "boolean"
+)
+
+const (
+	OrgAllowedScope  string = "org"
+	UserAllowedScope string = "user"
+)
+
+func (p *Preference) checkIfInAllowedValues(preferenceValue interface{}) bool {
+	isInAllowedValues := false
+	for _, value := range p.AllowedValues {
+		if value == preferenceValue {
+			isInAllowedValues = true
+		}
```

Comment:
this function is being called from the `isValidValue` check where the necessary type conversions have been made based on the `ValueType` of the preference

Reply:
While that may be true now, eventually you either forget that the passed value should be a comparable interface or someone who doesn't know the codebase touches this part and won't be familiar with the detail. We want to avoid that. You either make sure not to panic or explicitly call out that this expects comparable interfaces only.

Reply:
added checks in the `IsInAllowedValues` function to first check if the preference value is of the correct ValueType and then on each value of the allowedValues as well should be of the correct ValueType and then compare the same. if at any point something doesn't match with values return the error